### PR TITLE
perf(rpc): decode the block id object form in a single pass

### PR DESCRIPTION
**Pull request by brbrr**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Decode `BlockID` object JSON in one pass

- Apply same changes across RPC v9 and v10

- Add comprehensive unmarshal JSON tests

- Improve error handling for invalid inputs

___

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>block_id.go</strong><dd><code>Rewrite `BlockID` JSON unmarshal path</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rpc/v10/block_id.go

<ul><li>Add <code>bytes</code> import and typed <code>blockIDObject</code> struct<br> <li> Trim leading whitespace before choosing parse branch<br> <li> Decode object form in a single JSON pass<br> <li> Preserve tag handling and error behavior</ul>

</details>

  </td>
  <td><a href="https://github.com/NethermindEth/juno/pull/4009/files#diff-496f68fe1f2654b52086e3947cb4ae3ac7a3f286a72336517fc9b39e3649b61d">+36/-24</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>block.go</strong><dd><code>Rewrite `BlockID` JSON unmarshal path</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rpc/v9/block.go

<ul><li>Add <code>bytes</code> import and typed <code>blockIDObject</code> struct<br> <li> Trim leading whitespace before choosing parse branch<br> <li> Decode object form in a single JSON pass<br> <li> Preserve tag handling and error behavior</ul>

</details>

  </td>
  <td><a href="https://github.com/NethermindEth/juno/pull/4009/files#diff-3006d5a30c330e1ba9e1dcf358b810edad3cc50cfec46db77cda84b71f82138d">+36/-24</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>block_id_test.go</strong><dd><code>Add `BlockID` unmarshal JSON tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rpc/v10/block_id_test.go

<ul><li>Add table tests for tags, hash, and number<br> <li> Verify hash precedence and null fallback<br> <li> Include malformed, unknown, and raw error cases</ul>

</details>

  </td>
  <td><a href="https://github.com/NethermindEth/juno/pull/4009/files#diff-d1c14c1a7effd76c1a6e29792134bb5e086f42c1018f4a2ab43d9d1704221bfd">+87/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>block_id_test.go</strong><dd><code>Add `BlockID` unmarshal JSON tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rpc/v9/block_id_test.go

<ul><li>Add table tests for tags, hash, and number<br> <li> Verify hash precedence and null fallback<br> <li> Include malformed, unknown, and raw error cases</ul>

</details>

  </td>
  <td><a href="https://github.com/NethermindEth/juno/pull/4009/files#diff-ae95a0fd726be304f6f261bbcb7f9adaf0ab394b868b55d8fa0b9910208a04b4">+87/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: rpc/v10/block_id.go [hunk 1]
```diff
@@ -136,37 +137,48 @@ func (b *BlockID) Number() uint64 {
 	return b.data[0]
 }
 
+type blockIDObject struct {
+	BlockHash   *felt.Felt `json:"block_hash"`
+	BlockNumber *uint64    `json:"block_number"`
+}
```
Comment (rodrodros) [inline rpc/v10/block_id.go]: Please define this as part of `UnmarshalJSON`, since it is limited in scope to just that function

File: rpc/v9/block.go [hunk 2]
```diff
@@ -148,37 +149,48 @@ func (b *BlockID) Number() uint64 {
 	return b.data[0]
 }
 
+type blockIDObject struct {
+	BlockHash   *felt.Felt `json:"block_hash"`
+	BlockNumber *uint64    `json:"block_number"`
+}
```
Comment (rodrodros) [inline rpc/v9/block.go]: Similar thought

File: rpc/v10/block_id.go [hunk 3]
```diff
@@ -136,37 +137,48 @@ func (b *BlockID) Number() uint64 {
...
-			b.typeID = l1Accepted
-		default:
-			return fmt.Errorf("unknown block tag '%s'", blockTag)
-		}
-	} else {
-		jsonObject := make(map[string]json.RawMessage)
-		if err := json.Unmarshal(data, &jsonObject); err != nil {
+	trimmed := bytes.TrimLeft(data, " \t\r\n")
+	if len(trimmed) == 0 {
+		return errors.New("cannot unmarshal block id")
```
Comment (rodrodros) [inline rpc/v10/block_id.go:148]: Add test please (also in RPC v9)

**Review by rodrodros — APPROVED** [review]
Looks good, some nits
